### PR TITLE
Adds authorization for sidecar-promclient

### PR DESCRIPTION
**Pull request by someshkoli**
* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes
Added new flag `prometheus.http-client` to sidecar to pass http client config via file or string(similar to [alertmanager](https://thanos.io/tip/components/rule.md/#configuration)). Using this flag an instance of `thanoshttp.Client` is created and is passed to `promclient.NewClient`  method which uses this client as base client instead of whole new client which then makes authorized api calls.  
The same client is passed to reloader config which then makes authorized api calls to prom api.

There are some code duplication which has been duplicated from prometheus which can be easily removed by exposing one of the functions in prometheus `http_config`
:raised_hands: 
Close: #3975
## Verification
Added a e2e test:
Checks if query is able to query when prom uses auth

### CODE REVIEW

File: pkg/promclient/promclient.go [hunk 1]
```diff
@@ -800,3 +799,19 @@ func (c *Client) TargetsInGRPC(ctx context.Context, base *url.URL, stateTargets
...
+	MaxIdleConns        int
+	rt                  http.RoundTripper
+}
+
+func (r RoundTripperWithTransportOptions) RoundTrip(req *http.Request) (*http.Response, error) {
+	req2 := new(http.Request)
+	*req2 = *req
+	req2.Header = make(http.Header)
+	for k, s := range req.Header {
+		req2.Header[k] = s
```
Comment (someshkoli) [inline pkg/promclient/promclient.go:814]: Needs to be removed

File: cmd/thanos/sidecar.go [hunk 2]
```diff
@@ -212,10 +231,13 @@ func runSidecar(
 	}
 
 	{
+		// todo(someshkoli): Not sure how to properly implement RoundTripper for these properties or
```
Comment (someshkoli) [inline cmd/thanos/sidecar.go:234]: @Namanl2001 please have a look at this :thinking: I guess there's no clean way to set these properties. (might need to rewrite auth for promclient or make some changes thanoshttp for custom transporter)

Comment (wiardvanrij) [reply cmd/thanos/sidecar.go:234]: I think we need some 'generic' transporter that we can re-use everywhere. I had the same 'issue': https://github.com/thanos-io/thanos/pull/3970/files/e34e60e2848f2233809d67d5398151b670341ea1#diff-95c18d923c79bf32e4f8b3370d9a925aa910643559d351fab6bc307691c8f672R119 

The thanoshttp has one func which is:
```
// NewTransport creates a new http.Transport with default settings.
```
I guess we could add an other function that has httpConfig as argument for re-usability in the entire project.

Comment (someshkoli) [reply cmd/thanos/sidecar.go:234]: > I think we need some 'generic' transporter that we can re-use everywhere. I had the same 'issue': https://github.com/thanos-io/thanos/pull/3970/files/e34e60e2848f2233809d67d5398151b670341ea1#diff-95c18d923c79bf32e4f8b3370d9a925aa910643559d351fab6bc307691c8f672R119
> 
> The thanoshttp has one func which is:
> 
> ```
> // NewTransport creates a new http.Transport with default settings.
> ```
> 
> I guess we could add an other function that has httpConfig as argument for re-usability in the entire project.

Agreed, this is being used at 2-3 i guess and making a generic one can improve reusablity. @wiardvanrij I hope the overall idea in implementation is correct ? I can move forward with fixing this transporter issue ?

Comment (wiardvanrij) [reply cmd/thanos/sidecar.go:234]: I'm not the right person to make a definitive review. I miss the exact technical knowledge on it. However I was at the meeting and I think you made what was discussed. Also it looks logical and makes sense. 

You could always wait for a review if you think you might invest a lot of time into a feature you are unsure about. However if you have a good feeling about it, go make awesome stuff and continue 👍

Comment (someshkoli) [reply cmd/thanos/sidecar.go:234]: Alright thanks

File: pkg/promclient/promclient.go [hunk 3]
```diff
@@ -86,16 +86,16 @@ func NewClient(c HTTPClient, logger log.Logger, userAgent string) *Client {
...
 	return NewWithTracingClient(
 		log.NewNopLogger(),
+		http.Client{},
 		"",
 	)
 }
 
 // NewWithTracingClient returns client with tracing tripperware.
-func NewWithTracingClient(logger log.Logger, userAgent string) *Client {
+func NewWithTracingClient(logger log.Logger, httpClient http.Client, userAgent string) *Client {
```
Comment (bwplotka) [inline pkg/promclient/promclient.go:95]:
```suggestion
func NewWithTracingClient(logger log.Logger, httpClient *http.Client, userAgent string) *Client {
```

File: cmd/thanos/sidecar.go (hunk 2 above)
Comment (bwplotka) [reply cmd/thanos/sidecar.go:234]: I think this transport and connection flags 

```
conf.connection.maxIdleConnsPerHost
conf.connection.maxIdleConns
```

Are colliding with out `promethues.http-config` flag. The idea would be to unify those? E.g Deprecating connection flags if that makes sense? Or reusing them? 

The goal is to have one way of configuring HTTP configuration.

Comment (bwplotka) [reply cmd/thanos/sidecar.go:234]: Let's remove

>   --receive.connection-pool-size=RECEIVE.CONNECTION-POOL-SIZE
                                 Controls the http MaxIdleConns. Default is 0,
                                 which is unlimited
      --receive.connection-pool-size-per-host=100

And use our prometheus flags.

**Comment by someshkoli** [comment]
This implementation helps in providing a generic way to create http clients used across. There are still many instances where directly http library is used to create client which can be replaced with thanos http client instance and can be configured flexibly.
Also solves the issue mentined by @wiardvanrij  is [here](https://github.com/thanos-io/thanos/pull/4104#discussion_r620998741).

File: pkg/http/http.go [hunk 4]
```diff
@@ -63,6 +80,106 @@ func (b BasicAuth) IsZero() bool {
...
+	}
+
+	newRT := func(tlsConfig *tls.Config) (http.RoundTripper, error) {
+		var rt http.RoundTripper = &http.Transport{
+			Proxy:               http.ProxyURL(cfg.ProxyURL.URL),
+			MaxIdleConns:        opts.maxIdleConns,
+			MaxIdleConnsPerHost: opts.maxIdleConnsPerHost,
+			DisableKeepAlives:   true,
+			TLSClientConfig:     tlsConfig,
+			DisableCompression:  true,
```
Comment (wiardvanrij) [inline pkg/http/http.go:125]: I think we default to this because there was 'something with it' for Thanos side (_citation needed_). As this is more a generic package we should default to `false` and explicitly set this too `true` on functionalities via arguments that require this to be disabled. 

Tl;dr IMO:  set 'defaults' to something that would be defaults for _all_ packages, and override different cases when needed. Not the other way around. :)

File: pkg/http/http.go [hunk 5]
```diff
@@ -63,6 +80,106 @@ func (b BasicAuth) IsZero() bool {
...
+// NewRoundTripperFromConfig returns a new HTTP RoundTripper configured for the
+// given http.HTTPClientConfig and http.HTTPClientOption.
+func NewRoundTripperFromConfig(cfg config_util.HTTPClientConfig, name string, optFuncs ...HTTPClientOption) (http.RoundTripper, error) {
+	opts := defaultHTTPClientOptions
+	for _, f := range optFuncs {
+		f(&opts)
+	}
+
+	newRT := func(tlsConfig *tls.Config) (http.RoundTripper, error) {
+		var rt http.RoundTripper = &http.Transport{
```
Comment (wiardvanrij) [inline pkg/http/http.go:119]: This could be that I'm not looking correctly but aren't we missing certain options? Like `InsecureSkipVerify ` and `ResponseHeaderTimeout` ?

Comment (someshkoli) [reply pkg/http/http.go:119]: > This could be that I'm not looking correctly but aren't we missing certain options? Like `InsecureSkipVerify ` and `ResponseHeaderTimeout` ?

ResponseHeaderTimeout yes, but AFAIR `InsecureSkipVerify` is handled by tls ?

File: cmd/thanos/config.go [hunk 6]
```diff
@@ -69,24 +70,10 @@ func (pc *prometheusConfig) registerFlag(cmd extkingpin.FlagClause) *prometheusC
...
+	pc.httpClient = extkingpin.RegisterHTTPConfigFlags(cmd)
 	return pc
 }
 
-type connConfig struct {
-	maxIdleConns        int
-	maxIdleConnsPerHost int
-}
-
-func (cc *connConfig) registerFlag(cmd extkingpin.FlagClause) *connConfig {
```
Comment (bwplotka) [inline cmd/thanos/config.go:80]: Let's make sure we add this move to the changelog

File: pkg/http/http.go [hunk 7]
```diff
@@ -5,24 +5,38 @@
...
 	"github.com/prometheus/prometheus/discovery/file"
 	"github.com/prometheus/prometheus/discovery/targetgroup"
+	"golang.org/x/net/http2"
 	"gopkg.in/yaml.v2"
 
 	"github.com/thanos-io/thanos/pkg/discovery/cache"
 )
 
+var defaultHTTPClientOptions = httpClientOptions{
+	maxIdleConns:        0,
```
Comment (bwplotka) [inline pkg/http/http.go:36]: 0 feels wrong, see default Go impl:

```
// DefaultTransport is the default implementation of Transport and is
// used by DefaultClient. It establishes network connections as needed
// and caches them for reuse by subsequent calls. It uses HTTP proxies
// as directed by the $HTTP_PROXY and $NO_PROXY (or $http_proxy and
// $no_proxy) environment variables.
var DefaultTransport RoundTripper = &Transport{
	Proxy: ProxyFromEnvironment,
	DialContext: (&net.Dialer{
		Timeout:   30 * time.Second,
		KeepAlive: 30 * time.Second,
		DualStack: true,
	}).DialContext,
	ForceAttemptHTTP2:     true,
	MaxIdleConns:          100,
	IdleConnTimeout:       90 * time.Second,
	TLSHandshakeTimeout:   10 * time.Second,
	ExpectContinueTimeout: 1 * time.Second,
}

```

File: pkg/http/http.go [hunk 8]
```diff
@@ -5,24 +5,38 @@
...
+	maxIdleConnsPerHost: 100, // see https://github.com/golang/go/issues/13801
```
Comment (bwplotka) [inline pkg/http/http.go:37]:
```suggestion
	maxIdleConnsPerHost: 100, // See https://github.com/golang/go/issues/13801.
```

File: pkg/http/http.go (hunk 4 above)
Comment (bwplotka) [reply pkg/http/http.go:125]: We can, but we need to be aware who is using `exthttp.NewTransport` and how

File: pkg/http/http.go [hunk 9]
```diff
@@ -95,10 +212,23 @@ func NewHTTPClient(cfg ClientConfig, name string) (*http.Client, error) {
 		return nil, err
 	}
 
-	client, err := config_util.NewClientFromConfig(httpClientConfig, name, config_util.WithHTTP2Disabled())
+	// see https://github.com/golang/go/issues/13801
+	if cfg.MaxIdleConns == 0 {
+		cfg.MaxIdleConns = 20000
```
Comment (bwplotka) [inline pkg/http/http.go:217]: I don't get it, why we don't use default? Also `0` means no limit, so 0 has meaning too, we should tolerate such option

File: pkg/http/http.go [hunk 10]
```diff
@@ -95,10 +212,23 @@ func NewHTTPClient(cfg ClientConfig, name string) (*http.Client, error) {
 		return nil, err
 	}
 
-	client, err := config_util.NewClientFromConfig(httpClientConfig, name, config_util.WithHTTP2Disabled())
+	// see https://github.com/golang/go/issues/13801
```
Comment (bwplotka) [inline pkg/http/http.go:215]: Weird haging comment (and not full sentence), can we remove it?

File: pkg/http/http.go [hunk 11]
```diff
@@ -95,10 +212,23 @@ func NewHTTPClient(cfg ClientConfig, name string) (*http.Client, error) {
...
+	if cfg.MaxIdleConns == 0 {
+		cfg.MaxIdleConns = 20000
+	}
+	if cfg.MaxIdleConnsPerHost == 0 {
+		cfg.MaxIdleConnsPerHost = 1000
```
Comment (bwplotka) [inline pkg/http/http.go:220]: Why those numbers?

File: pkg/http/http.go [hunk 12]
```diff
@@ -95,10 +212,23 @@ func NewHTTPClient(cfg ClientConfig, name string) (*http.Client, error) {
...
+	}
+	rt, err := NewRoundTripperFromConfig(
+		httpClientConfig,
+		name,
+		WithMaxIdleConns(cfg.MaxIdleConns),
+		WithMaxIdleConnsPerHost(cfg.MaxIdleConnsPerHost))
 	if err != nil {
 		return nil, err
 	}
+	client := &http.Client{Transport: rt}
```
Comment (bwplotka) [inline pkg/http/http.go:230]: Setting transport does not make sense, if we override it later on, why not passing `rt` just below?

File: pkg/http/http.go [hunk 13]
```diff
@@ -262,3 +392,151 @@ func (c *Client) Discover(ctx context.Context) {
 func (c *Client) Resolve(ctx context.Context) error {
 	return c.provider.Resolve(ctx, append(c.fileSDCache.Addresses(), c.staticAddresses...))
 }
+
+// Duplicated from github.com/prometheus/common/config
+// better option would be to expose config_util.newTLSRoundTripper function from prometheus
```
Comment (bwplotka) [inline pkg/http/http.go:397]: Yes, can we do that (you can contribute there, we are happy to review and merge it there)

File: cmd/thanos/config.go (hunk 6 above)
Comment (bwplotka) [reply cmd/thanos/config.go:80]: Why it was receive? 🤔

**Review by bwplotka — COMMENTED** [review]
Don't have time to review fully, but did part of it, looks good generally, some nits.

File: pkg/http/http.go (hunk 7 above)
Comment (someshkoli) [reply pkg/http/http.go:36]: > 0 feels wrong, see default Go impl:
> 
> ```
> // DefaultTransport is the default implementation of Transport and is
> // used by DefaultClient. It establishes network connections as needed
> // and caches them for reuse by subsequent calls. It uses HTTP proxies
> // as directed by the $HTTP_PROXY and $NO_PROXY (or $http_proxy and
> // $no_proxy) environment variables.
> var DefaultTransport RoundTripper = &Transport{
> 	Proxy: ProxyFromEnvironment,
> 	DialContext: (&net.Dialer{
> 		Timeout:   30 * time.Second,
> 		KeepAlive: 30 * time.Second,
> 		DualStack: true,
> 	}).DialContext,
> 	ForceAttemptHTTP2:     true,
> 	MaxIdleConns:          100,
> 	IdleConnTimeout:       90 * time.Second,
> 	TLSHandshakeTimeout:   10 * time.Second,
> 	ExpectContinueTimeout: 1 * time.Second,
> }
> ```

Yes right

File: pkg/http/http.go (hunk 12 above)
Comment (someshkoli) [reply pkg/http/http.go:230]: Makes sense

File: pkg/http/http.go (hunk 11 above)
Comment (someshkoli) [reply pkg/http/http.go:220]: Again from prometheus implementation.

File: pkg/http/http.go (hunk 13 above)
Comment (someshkoli) [reply pkg/http/http.go:397]: Great, I'll raise a PR there

File: pkg/http/http.go (hunk 9 above)
Comment (someshkoli) [reply pkg/http/http.go:217]: > I don't get it, why we don't use default? Also `0` means no limit, so 0 has meaning too, we should tolerate such option

Actually, this value is picked from `prometheus http` which we were using before, so thought it would be better to keep it the same. Also yeah 0 does for this is acceptable, 'll fix it.

File: cmd/thanos/config.go (hunk 6 above)
Comment (someshkoli) [reply cmd/thanos/config.go:80]: By mistake I assume :thinking: I don't see it being used anywhere

File: pkg/http/http.go (hunk 13 above)
Comment (someshkoli) [reply pkg/http/http.go:392]: https://github.com/prometheus/common/pull/301

File: cmd/thanos/sidecar.go [hunk 14]
```diff
@@ -84,7 +81,24 @@ func runSidecar(
...
+		return errors.Wrap(err, "getting http client config")
+	}
+	httpClientConfig, err := thanoshttp.NewClientConfigFromYAML(httpConfContentYaml)
+	if err != nil {
+		return errors.Wrap(err, "parsing http config YAML")
+	}
+
+	httpClient, err := thanoshttp.NewHTTPClient(*httpClientConfig, "thanos-sidecar")
+	if err != nil {
+		return errors.Wrap(err, "Improper http client config")
```
Comment (bwplotka) [inline cmd/thanos/sidecar.go:95]:
```suggestion
		return errors.Wrap(err, "new http client config")
```

File: pkg/http/http.go [hunk 15]
```diff
@@ -35,6 +39,8 @@ type ClientConfig struct {
 	ProxyURL string `yaml:"proxy_url"`
 	// TLSConfig to use to connect to the targets.
 	TLSConfig TLSConfig `yaml:"tls_config"`
+	// TransportConfig for Client transport properties
```
Comment (bwplotka) [inline pkg/http/http.go:42]:
```suggestion
	// TransportConfig for Client transport properties.
```

File: pkg/http/http.go [hunk 16]
```diff
@@ -91,15 +195,31 @@ func NewHTTPClient(cfg ClientConfig, name string) (*http.Client, error) {
...
+
+	if cfg.BearerTokenFile != "" {
+		httpClientConfig.BearerTokenFile = cfg.BearerTokenFile
+	}
+
 	if err := httpClientConfig.Validate(); err != nil {
 		return nil, err
 	}
 
-	client, err := config_util.NewClientFromConfig(httpClientConfig, name, config_util.WithHTTP2Disabled())
```
Comment (bwplotka) [inline pkg/http/http.go:98]: Can we still use it? and append middleware we want on top?

**Review by bwplotka — COMMENTED** [review]
LGTM generally, just small nits. Thanks.

**Comment by bwplotka** [comment]
Any update on this?

**Comment by someshkoli** [comment]
> Any update on this?

Its ready for a review, I guess I forgot to put a review request :smile:

**Comment by yeya24** [comment]
@someshkoli Can you rebase on main and fix the conflicts? Thank you!

**Comment by yeya24** [comment]
Hi, @someshkoli, sorry for the late reply. Can you fix the conflict and get CI passing? Your work is really amazing and we would like to have this feature!

**Comment by someshkoli** [comment]
> Hi, @someshkoli, sorry for the late reply. Can you fix the conflict and get CI passing? Your work is really amazing and we really want to have this feature!

will update it soon, have a lot to rebase :")

**Comment by someshkoli** [comment]
Moving this to #4612

**Comment by someshkoli** [comment]
@yeya24 can you please take a look at #4612 ?

**Comment by yeya24** [comment]
Close in favor of #4612
